<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_13) on Fri Mar 28 12:37:37 CDT 2008 -->
<TITLE>
RefGenomeServiceSleep
</TITLE>

<META NAME="keywords" CONTENT="org.bbop.client.RefGenomeServiceSleep interface">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="RefGenomeServiceSleep";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/RefGenomeServiceSleep.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/bbop/client/RefGenomeServiceClientImpl.html" title="class in org.bbop.client"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/bbop/client/WebSession.html" title="class in org.bbop.client"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/bbop/client/RefGenomeServiceSleep.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RefGenomeServiceSleep.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.bbop.client</FONT>
<BR>
Interface RefGenomeServiceSleep</H2>
<DL>
<DT><B>All Superinterfaces:</B> <DD>com.google.gwt.user.client.rpc.RemoteService</DD>
</DL>
<HR>
<DL>
<DT><PRE>public interface <B>RefGenomeServiceSleep</B><DT>extends com.google.gwt.user.client.rpc.RemoteService</DL>
</PRE>

<P>
<HR>

<P>

<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#addUser(java.lang.String, java.lang.String, java.lang.String)">addUser</A></B>(java.lang.String&nbsp;userId,
        java.lang.String&nbsp;fullName,
        java.lang.String&nbsp;password)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;given an entity Id (e.g.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#assignHomologyLinkStatement(java.lang.String, java.lang.String, java.lang.String, java.lang.String[], java.lang.String, java.lang.String)">assignHomologyLinkStatement</A></B>(java.lang.String&nbsp;userId,
                            java.lang.String&nbsp;e1id,
                            java.lang.String&nbsp;e2id,
                            java.lang.String[]&nbsp;methodIds,
                            java.lang.String&nbsp;provenanceId,
                            java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;e.g.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#assignNegativeHomologyLinkStatement(java.lang.String, java.lang.String, java.lang.String, java.lang.String[], java.lang.String, java.lang.String)">assignNegativeHomologyLinkStatement</A></B>(java.lang.String&nbsp;userId,
                                    java.lang.String&nbsp;e1id,
                                    java.lang.String&nbsp;e2id,
                                    java.lang.String[]&nbsp;methodIds,
                                    java.lang.String&nbsp;provenanceId,
                                    java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#attachCommentToEntity(java.lang.String, java.lang.String, java.lang.String)">attachCommentToEntity</A></B>(java.lang.String&nbsp;entityId,
                      java.lang.String&nbsp;comment,
                      java.lang.String&nbsp;curatorId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#checkUserPassword(java.lang.String, java.lang.String)">checkUserPassword</A></B>(java.lang.String&nbsp;userId,
                  java.lang.String&nbsp;password)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchEntityIdsInHomologSet(java.lang.String)">fetchEntityIdsInHomologSet</A></B>(java.lang.String&nbsp;homologSetId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given an ID for the homology set, return the IDs of all members</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchHomologousEntityIds(java.lang.String)">fetchHomologousEntityIds</A></B>(java.lang.String&nbsp;entityId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;given a gene (or similar entity) return all homologs</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchIdsByName(java.lang.String)">fetchIdsByName</A></B>(java.lang.String&nbsp;searchTerm)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A Facade for accessing a RefG repository
 
 All IDs are assumed to uniquely identify the object of interest, and
 are of the form DB : LocalID</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchIdsByNameAndTaxon(java.lang.String, java.lang.String)">fetchIdsByNameAndTaxon</A></B>(java.lang.String&nbsp;searchTerm,
                       java.lang.String&nbsp;taxonId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;limits search results by taxon</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Map</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchLabelMapsById(java.lang.String)">fetchLabelMapsById</A></B>(java.lang.String&nbsp;searchTerm)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchLabelsById(java.lang.String)">fetchLabelsById</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given an Id, return all labels, names, synonyms etc by which this Id is
 known</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchReferenceTargetIds()">fetchReferenceTargetIds</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#fetchReferenceTaxonIds()">fetchReferenceTaxonIds</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;all taxon IDs that are part of some reference set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#getTaxonIdPrefix()">getTaxonIdPrefix</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;typical NCBITaxon</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#retractEntityComprehensivelyAnnotatedStatus(java.lang.String, java.lang.String)">retractEntityComprehensivelyAnnotatedStatus</A></B>(java.lang.String&nbsp;userId,
                                            java.lang.String&nbsp;geneId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#retractEntityStatusCode(java.lang.String, java.lang.String, java.lang.String)">retractEntityStatusCode</A></B>(java.lang.String&nbsp;userId,
                        java.lang.String&nbsp;statusId,
                        java.lang.String&nbsp;geneId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/bbop/client/RefGenomeServiceSleep.html#retractEntityTargetStatus(java.lang.String, java.lang.String)">retractEntityTargetStatus</A></B>(java.lang.String&nbsp;userId,
                          java.lang.String&nbsp;geneId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="fetchIdsByName(java.lang.String)"><!-- --></A><H3>
fetchIdsByName</H3>
<PRE>
java.lang.String[] <B>fetchIdsByName</B>(java.lang.String&nbsp;searchTerm)</PRE>
<DL>
<DD>A Facade for accessing a RefG repository
 
 All IDs are assumed to uniquely identify the object of interest, and
 are of the form DB : LocalID
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>searchTerm</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="fetchIdsByNameAndTaxon(java.lang.String, java.lang.String)"><!-- --></A><H3>
fetchIdsByNameAndTaxon</H3>
<PRE>
java.lang.String[] <B>fetchIdsByNameAndTaxon</B>(java.lang.String&nbsp;searchTerm,
                                          java.lang.String&nbsp;taxonId)</PRE>
<DL>
<DD>limits search results by taxon
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>searchTerm</CODE> - -- eg SOX9<DD><CODE>taxonId</CODE> - -- eg NCBITax:7227
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="fetchLabelsById(java.lang.String)"><!-- --></A><H3>
fetchLabelsById</H3>
<PRE>
java.lang.String[] <B>fetchLabelsById</B>(java.lang.String&nbsp;id)</PRE>
<DL>
<DD>Given an Id, return all labels, names, synonyms etc by which this Id is
 known
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>id</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="fetchLabelMapsById(java.lang.String)"><!-- --></A><H3>
fetchLabelMapsById</H3>
<PRE>
java.util.Map <B>fetchLabelMapsById</B>(java.lang.String&nbsp;searchTerm)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="fetchReferenceTaxonIds()"><!-- --></A><H3>
fetchReferenceTaxonIds</H3>
<PRE>
java.lang.String[] <B>fetchReferenceTaxonIds</B>()</PRE>
<DL>
<DD>all taxon IDs that are part of some reference set.
 the API is extensible to multiple reference sets - this call
 will always return the superset. In the case of the RefG
 project, there should always be 12 IDs returned
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>IDs for taxon nodes, typically NCBITaxon</DL>
</DD>
</DL>
<HR>

<A NAME="getTaxonIdPrefix()"><!-- --></A><H3>
getTaxonIdPrefix</H3>
<PRE>
java.lang.String <B>getTaxonIdPrefix</B>()</PRE>
<DL>
<DD>typical NCBITaxon
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>IDSpace of the taxononymy database</DL>
</DD>
</DL>
<HR>

<A NAME="fetchReferenceTargetIds()"><!-- --></A><H3>
fetchReferenceTargetIds</H3>
<PRE>
java.lang.String[] <B>fetchReferenceTargetIds</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>all target gene Ids</DL>
</DD>
</DL>
<HR>

<A NAME="fetchEntityIdsInHomologSet(java.lang.String)"><!-- --></A><H3>
fetchEntityIdsInHomologSet</H3>
<PRE>
java.lang.String[] <B>fetchEntityIdsInHomologSet</B>(java.lang.String&nbsp;homologSetId)</PRE>
<DL>
<DD>Given an ID for the homology set, return the IDs of all members
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>homologSetId</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="fetchHomologousEntityIds(java.lang.String)"><!-- --></A><H3>
fetchHomologousEntityIds</H3>
<PRE>
java.lang.String[] <B>fetchHomologousEntityIds</B>(java.lang.String&nbsp;entityId)</PRE>
<DL>
<DD>given a gene (or similar entity) return all homologs
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>entityId</CODE> - -- e.g. a gene or gene product Id
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="addUser(java.lang.String, java.lang.String, java.lang.String)"><!-- --></A><H3>
addUser</H3>
<PRE>
void <B>addUser</B>(java.lang.String&nbsp;userId,
             java.lang.String&nbsp;fullName,
             java.lang.String&nbsp;password)</PRE>
<DL>
<DD>given an entity Id (e.g. a MOD gene ID), return all edges in
 the homology graph
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>entityId</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="checkUserPassword(java.lang.String, java.lang.String)"><!-- --></A><H3>
checkUserPassword</H3>
<PRE>
boolean <B>checkUserPassword</B>(java.lang.String&nbsp;userId,
                          java.lang.String&nbsp;password)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="retractEntityStatusCode(java.lang.String, java.lang.String, java.lang.String)"><!-- --></A><H3>
retractEntityStatusCode</H3>
<PRE>
void <B>retractEntityStatusCode</B>(java.lang.String&nbsp;userId,
                             java.lang.String&nbsp;statusId,
                             java.lang.String&nbsp;geneId)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="retractEntityTargetStatus(java.lang.String, java.lang.String)"><!-- --></A><H3>
retractEntityTargetStatus</H3>
<PRE>
void <B>retractEntityTargetStatus</B>(java.lang.String&nbsp;userId,
                               java.lang.String&nbsp;geneId)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="retractEntityComprehensivelyAnnotatedStatus(java.lang.String, java.lang.String)"><!-- --></A><H3>
retractEntityComprehensivelyAnnotatedStatus</H3>
<PRE>
void <B>retractEntityComprehensivelyAnnotatedStatus</B>(java.lang.String&nbsp;userId,
                                                 java.lang.String&nbsp;geneId)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="attachCommentToEntity(java.lang.String, java.lang.String, java.lang.String)"><!-- --></A><H3>
attachCommentToEntity</H3>
<PRE>
void <B>attachCommentToEntity</B>(java.lang.String&nbsp;entityId,
                           java.lang.String&nbsp;comment,
                           java.lang.String&nbsp;curatorId)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="assignHomologyLinkStatement(java.lang.String, java.lang.String, java.lang.String, java.lang.String[], java.lang.String, java.lang.String)"><!-- --></A><H3>
assignHomologyLinkStatement</H3>
<PRE>
void <B>assignHomologyLinkStatement</B>(java.lang.String&nbsp;userId,
                                 java.lang.String&nbsp;e1id,
                                 java.lang.String&nbsp;e2id,
                                 java.lang.String[]&nbsp;methodIds,
                                 java.lang.String&nbsp;provenanceId,
                                 java.lang.String&nbsp;comment)</PRE>
<DL>
<DD>e.g. PPOD, Homologene, ...
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="assignNegativeHomologyLinkStatement(java.lang.String, java.lang.String, java.lang.String, java.lang.String[], java.lang.String, java.lang.String)"><!-- --></A><H3>
assignNegativeHomologyLinkStatement</H3>
<PRE>
void <B>assignNegativeHomologyLinkStatement</B>(java.lang.String&nbsp;userId,
                                         java.lang.String&nbsp;e1id,
                                         java.lang.String&nbsp;e2id,
                                         java.lang.String[]&nbsp;methodIds,
                                         java.lang.String&nbsp;provenanceId,
                                         java.lang.String&nbsp;comment)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/RefGenomeServiceSleep.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/bbop/client/RefGenomeServiceClientImpl.html" title="class in org.bbop.client"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/bbop/client/WebSession.html" title="class in org.bbop.client"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/bbop/client/RefGenomeServiceSleep.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RefGenomeServiceSleep.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
